### DISCUSSION TITLE
CSS Tailwind Syntax

### DISCUSSION THREAD

**Discussion by lishaduck** [biomejs/biome#3195] [Ideas]
> [!TIP]
> There is now an issue, please (don't?) discuss there https://github.com/biomejs/biome/issues/7223
> Biome 2.3 will include the associated patch, so all we've got left to do is wait.
> For now, you can use https://pkg.pr.new/@biomejs/biome@72a8b02.

> [!IMPORTANT]
> _PLEASE THINK BEFORE COMMENTING_. Biome is an OSS project, and things don't change by continually harassing maintainers with +1-style comments.
> If this is an actual issue for your company, please come implement it yourself. If this is just an issue for you personally, I'm sorry. I do feel you. I hate to ignore rules, but sometimes it's a necessary evil.
> — OP 

---

Update: Biome now seems to parse Tailwind 3 properly, but there are reports below of it failing on Tailwind 4's syntax.

---

Tailwind adds a few custom at-rules. I tried to see how well the new css work does on my css, but it fails at my `@apply` declarations. I can comment them out, but that's not a long-term solution. Probably the parser should just parse `@apply`. Surprisingly, `@tailwind` just works. Eventually, I'd like it if #1274 could also be integrated into `@apply`, but that's certainly not needed at this point.

Thanks for considering this! Biome is great!

P.S. After commenting it out, it linted my css just fine! It only found an empty ruleset b/c I'd commented it out, but it's running, finding issue, and, of course, it's fast ;)

**Comment by znycheporuk** [comment]
Could you clarify if Biome will support the Tailwind v4 syntax, which includes the following to prevent the generation of default CSS variables, or if it will only support standard CSS syntax? I'm ready to contribute if you are open to supporting it.

```css
@import "tailwindcss";

@theme {
	--color-*: initial;
	--spacing-*: initial;
}
```

<img width="629" alt="image" src="https://github.com/user-attachments/assets/738fbbaf-9534-4ad0-9250-c8549e38d40a">

**Reply by znycheporuk** [reply]
this is already fixed on tailwind side 
https://github.com/tailwindlabs/tailwindcss/pull/15603

**Comment by mkvlrn** [comment]
Hijacking so there isn't yet another thread about this.  Trying v4 today and [setting a base path for tailwind using the source function](https://tailwindcss.com/docs/detecting-classes-in-source-files#setting-your-base-path) with @import is an error:

![image](https://github.com/user-attachments/assets/3fd8d134-32ad-4747-89a8-26349c0898c8)
![image](https://github.com/user-attachments/assets/ff7ef5a0-a58e-4170-93e6-f3e8e5de5829)

**Reply by kisamoto** [reply]
As a workaround to this I moved the source to:

```
@import "tailwindcss";
@source ".";
```

**Comment by xcrap** [comment]
<img width="538" alt="Screenshot 2025-02-03 at 13 54 11" src="https://github.com/user-attachments/assets/e87ba50d-59d3-42a4-bb3f-e4f7af163fd3" />
<img width="470" alt="Screenshot 2025-02-03 at 13 54 16" src="https://github.com/user-attachments/assets/fc99e4a1-c304-4560-a3c5-4270e137b172" />

Yeah the work around does not work with source(none); and even using biome-ignore it stills warns me about this, any solution ?

**Reply by ekwoka** [reply]
This is because it's not a linting error, but a parsing error...

**Comment by schang124** [comment]
Seems like there is no workaround for [prefix for tailwind v4](https://tailwindcss.com/docs/upgrade-guide#using-a-prefix).
If Biome support tailwind v4 syntax, that would be great since lots of people uses it.
```
@import "tailwindcss" prefix(tw); // <-- this gives error
```
![image](https://github.com/user-attachments/assets/9b48c924-52d9-479b-bb69-a99e088c9ef4)

**Comment by nezzard** [comment]
Any updates?

**Comment by jole78** [comment]
might be totally wrong here but I think this works...does for me at least

```
@import 'tailwindcss/theme' layer(theme);
@import 'tailwindcss/utilities' layer(utilities);

@theme prefix(tw) {
  --\*: initial;
  ...your theme here...
```

**Comment by EKliptus** [comment]
In my case, Biome can't format css files when the necessary Vscode settings associate css file to tailwind. Biome isn't associate to tailwind languageID :
```jsonc
// settings.json
 "files.associations": {
    "*.css": "tailwindcss" // https://marketplace.visualstudio.com/items?itemName=bradlc.vscode-tailwindcss#files.associations
  }
```
It is needed for vscode to recognize `@plugin` directive and others

With Biome is it possible to override file extension to tell all css file is parse/format/lint/check by Biome  ?

**Reply by lishaduck** [reply]
That's a different issue, this is about syntax support (although if you fixed it, then you'd probably hit this as well)

**Comment by therealJonSnow** [comment]
I came across the same issue when using `--color-*: initial;` as per the tailwind docs for removing all tailwind default colors.

@jole78 solution seems to remove all tailwind default classes (including margin, padding etc).

If you just want to remove colors then i found the `--color-\*: initial;` worked.

I have literally no experience with biomejs - but hope this helps and works for others.

**Reply by jesseb34r** [reply]
escaping with a `\` also worked for me

**Reply by AlbaOngaro** [reply]
same here!

**Comment by theskillwithin** [comment]
still not fixed in biome v2

**Comment by Mnigos** [comment]
Any workaround for this?

**Reply by ekwoka** [reply]
escaping with `\*` works fine.

**Comment by cantutar** [comment]
My problem was it picks up @apply and @plugin and @custom-variant  for example: 
```bash
src/styles/index.css:4:2 lint/suspicious/noUnknownAtRules ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ✖ Unexpected unknown at-rule: plugin 
  
    2 │ @import "tw-animate-css";
    3 │ /* index.css veya global CSS'te */
  > 4 │ @plugin "tailwindcss-debug-screens";
      │  ^^^^^^
    5 │ 
```

**Reply by yairEO** [reply]
Yeah.. really terrible thing. I would prefer Biome to allow the `biome.json` file to "extend" the `noUnknownAtRules` with a whitelist, something like so:

```
{
  "linter": {
    "rules": {
      "suspicious": {
        "noUnknownAtRules": {
          "level": "warn",
          "options": {
            "whitelist": ["@apply", "@plugin"]
          }
        }  
      }
    }
  }
}
```

**Reply by Netail** [reply]
Stylelint also has an exclude option; https://stylelint.io/user-guide/rules/at-rule-no-unknown/. I think this would be nice to add, even if Tailwind syntax will be supported soon

**Reply by rickyzhangca** [reply]
getting `Property whitelist is not allowed.` for using `whitelist`

**Reply by lishaduck** [reply]
> getting `Property whitelist is not allowed.` for using `whitelist`

Please read the whole comment before responding. That was an idea for what might be added someday if someone decides to implement generic CSS rule support. As is, **there is no such support**.

**Reply by Netail** [reply]
Biome 2.3.0 will land an ignore option; https://github.com/biomejs/biome/pull/7745. But will also support tailwind syntax, so probably won't be needed. More a future proof feature ;)

**Reply by smartcraze** [reply]
use the this 
"linter": {
    "domains": {
      "next": "recommended",
      "react": "recommended"
    },
    "rules": {
      "suspicious": {
        "noUnknownAtRules": {
          "level": "warn",
          "options": {
            "ignore": ["@apply", "@plugin", "@theme"]
          }
        }
      }
    }
  }

**Comment by abcfy2** [comment]
Any update?

**Comment by ematipico** [comment]
This has been a bit lowkey, but we shipped support for Tailwind CSS files, and it will be available in v2.3: https://github.com/biomejs/biome/pull/7164

Great thanks for @dyc3 for that! 

We will enable the feature in the playground soon :)

**Comment by PaulRBerg** [comment]
I've enabled the CSS parser in Biome v2.3, but I am still getting an error when linting this `variants.css` file:

```css
@custom-variant max-xxs (@media (max-width: 320px));
```

Getting this parse error:

> Expected a selector but instead found '@'.biome(parse)

Should I open an issue?

**Reply by ematipico** [reply]
Yes please. Provide a playground reproduction and a link to the tailwind docs that shows that

**Reply by Netail** [reply]
> Should I open an issue?

If you could, that would be lovely. Don't see any issues open related to this directive yet

**Reply by PaulRBerg** [reply]
https://github.com/biomejs/biome/issues/7964

**Comment by ghost** [comment]
No support for prefix

<img width="896" height="367" alt="image" src="https://github.com/user-attachments/assets/67488535-3c05-47e4-a359-8b200491bcc9" />

**Reply by dyc3** [reply]
see #7920

**Reply by ghost** [reply]
> see #7920

Is there a way to fully ignore the error until it is fixed?

**Reply by dyc3** [reply]
No, not at the moment.
